### PR TITLE
Upgrade LEAN Python to 3.6 from 2.7

### PR DESCRIPTION
**Pull request by mchandschuh**
Slightly reworked version of https://github.com/QuantConnect/Lean/pull/1526

The biggest change from #1526 is that we're now installing miniconda into `/opt/` so it's available to all users.

---
#### Description from #1526 
In this pull request we propose changes that enables python 3.6 in Lean.
- Adds log to display the python version the algorithm is using.
- Fixes python algorithms that were failing because of small subtleties like leading zeroes.
- Updates pythonnet with a version compiled with python 3.6 flags

Also included via #1526:
- Changes in DockerfileFoundation: we now use miniconda to manage the python environment. Took the opportunity to add NTLK (#1349), Tensorforce (#1369) and PyTorch/Pyro (#1385).
- Changes readme in Algorithm.Python to show steps to install miniconda

### CODE REVIEW

File: DockerfileLeanFoundation [hunk 1]
```diff
@@ -14,35 +14,39 @@ CMD ["/sbin/my_init"]
...
-  pip install blaze cvxopt cvxpy pykalman statistics statsmodels arch copulalib && \
-  pip install xgboost theano keras tensorflow deap
+RUN apt-get update && apt-get install -y wget bzip2 python3-pip python-opengl && \
+    apt-get install -y xvfb unzip curl libxrender1 libxtst6 libxi6
+
+# Install miniconda and supported third party python packages
+ENV PATH="/opt/miniconda3/bin:${PATH}"
+RUN wget https://repo.continuum.io/miniconda/Miniconda3-4.3.31-Linux-x86_64.sh && \
+    bash Miniconda3-4.3.31-Linux-x86_64.sh -b -p /opt/miniconda3 && \
+    rm -rf Miniconda3-latest-Linux-x86_64.sh && \
```
Comment (AlexCatarino) [inline DockerfileLeanFoundation]: I think we should use:
```
RUN wget https://repo.continuum.io/miniconda/Miniconda3-4.3.31-Linux-x86_64.sh -O miniconda.sh && \
    bash miniconda.sh -b -p /opt/miniconda3 && \
    rm -rf miniconda.sh && \
```
To keep the version in only one line. In this case we were deleting a non-existing file.

File: Algorithm.Python/readme.md [hunk 2]
```diff
@@ -113,15 +113,28 @@ $ sudo easy-install pip
 ### Linux
 **1. Follow the installation instructions [here](https://github.com/QuantConnect/Lean#linux-debian-ubuntu)**
 
-1.1 You should see this exit successfully with a list of the algorithm statistics and 1 trade.
+**2. Install Python 3.6 with Miniconda:**
+```
+export PATH="/root/miniconda3/bin:$PATH"
+wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh
+bash Miniconda3-latest-Linux-x86_64.sh -b
+rm -rf Miniconda3-latest-Linux-x86_64.sh
```
Comment (AlexCatarino) [inline Algorithm.Python/readme.md]: Here we should have the same commands of the dockerfile.

File: DockerfileLeanFoundation (hunk 1 above)
Comment (mchandschuh) [reply DockerfileLeanFoundation]: I actually did something similar you shared with me, but when I ran it, docker couldn't find the renamed script, so I dd it this way instead.

**Comment by disruptek** [comment]
Some more missing items to add to the docs:

```python
# implement cmp()
cmp = lambda x, y: (x > y)-(x < y)

# implement basestring
try:
    basestring
except NameError:
    basestring = str
```

I had a few others but every time you guys spring this upgrade upon me, I find and fix a few more.

**Pull request merged**